File: Src/led.c
#include "led.h"
#include "gpio.h"

void led_init() {
    led_off(LED1);
    led_off(LED2);
}

void led_on(LedInfo led) {
    atomic_gpio_clear_pin_output(led.port, led.pin);
}

void led_off(LedInfo led) {
    atomic_gpio_set_pin_output(led.port, led.pin);
}

void led_toggle(LedInfo led) {
    PinState pin_state = gpio_pin_status(led.port, led.pin);
    pin_state ? led_off(led) : led_on(led);
}

File: zig-src/led.zig
const af = @import("af.zig");
const gpio = @import("gpio.zig");

pub const Led = struct {
    pin: gpio.PinConfig,

    pub fn on(self: Led) void {
        gpio.clear(self.pin);
    }

    pub fn off(self: Led) void {
        gpio.set(self.pin);
    }

    pub fn toggle(self: Led) void {
        if (gpio.read(self.pin) == .high) {
            self.on();
        } else {
            self.off();
        }
    }
};

pub const led1 = Led{ .pin = .{
    .port = .g,
    .pin = .p9,
    .mode = .output,
    .output = .push_pull,
    .pull = .down,
    .alternate = af.none,
} };

pub const led2 = Led{ .pin = .{
    .port = .f,
    .pin = .p11,
    .mode = .output,
    .output = .push_pull,
    .pull = .down,
    .alternate = af.none,
} };

pub fn init() void {
    gpio.configure(led1.pin);
    gpio.configure(led2.pin);
    led1.off();
    led2.on();
}
